### ISSUE TITLE
have you looked at hxPixels ?

### ISSUE THREAD

**Issue by nanjizal** [cancerberoSgx/bitmap#2]
Have you looked at hxPixels, perhaps there is some overlap.

**Comment by cancerberoSgx** [comment]
Thanks, There is some overlap. But I see some differences too. Probably both projects could copy things from each other.

I'm mainly working on this to practice with haxe since I'm new. Also I want to load pngs and have basic image processing, without depending on lime/openfl and supporting all targets, including js and browser, using the same API. This objective I already accomplished and now I'm adding some transformations like convolutions/affine/color filters. And also a basic command line interface. 

For this project main objectives I think I'm done, again did this mostly to practice with the programming language. Nevertheless I will take a more detailed look to that project , probably I can contribute with several things and viceversa. Thanks!

**Comment by nanjizal** [comment]
Perhaps worth separating loading and processing images?
I have this repo for loading images a wrapper for format library and makes use of hxPixels:
https://github.com/nanjizal/folder
You can see it's use with 
https://github.com/nanjizal/hxRectPack2D_pixel
For nodejs slightly different approach
https://github.com/nanjizal/folderNode
You can see it's use within my shopify externs on github.
Perhaps irrelevant but perhaps useful as alternate approaches.  
hxPixels is used with quite a few of my libraries and was developed by Azarfe7 to provide pixel access for the hxDaedalus library we ported, in trilateralBazzar or mandelbrot you can find hxPixels use with Kha.

**Comment by cancerberoSgx** [comment]
> Perhaps worth separating loading and processing images?

yes that would make sense. As said, I'm just getting fun with haxe exploring the language and libraries. Probably I will split this in separate projects (encoder/decoder, basic API and resource access, CLI, high level API, plugins for transformations, filters, etc.) . Right now I have it all mixed in the same place. I love learning new languages and passionate about image processing so I will take care of that later (BTW there is a simple browser demo (not using canvas, and encoding/decoding each time https://cancerberosgx.github.io/bitmap/playground/ but nevertheless is quite fast). WIll try to make it a real playground to testing with more combinations.

It would be nice if we can exchange experience or things we have to complement each other. here some thoughts:

CONS:
 * I couldn't find a way of read jpeg. I see you have some support in your project but say it's buggy and licencing trouble? I think not having a haxe library for reading JPEG at least in my target audience is a big problem. An idea I have on this is not haxe but it could help and support many formats (and I didn't see that implemented in haxelib). Server side use imagemagick to convert to png. Format conversion ispeed is not critical since it happens only once. And on the browser (and even on node) use a imagemagick emscripten port like https://www.npmjs.com/package/magica. This way we can support format conversion (and any image magick CLI) in all targets. THe downside is that for targets different than js you need IM installed locally. I think that will be my next project if any. An ultimately objective for me is, if the lib is 100% haxe , to export it to native js, cpp, java libraries automatically. Is not that it will be useful since those have many, but I want to have that experience. 

Regarding this my first problem was to make PNG encoding to work in js, since haxe.zip doesn't really  support it (https://github.com/HaxeFoundation/haxe/issues/8540) - there description of the workaround.  Now I have the same code running on all targets with some native javascript (zlib) that is injected for js.I have plenty ideas and taking advantage of free time


> You can see it's use within my shopify externs on github.
> Perhaps irrelevant but perhaps useful as alternate approaches.
> hxPixels is used with quite a few of my libraries and was developed by Azarfe7 to provide pixel access for the hxDaedalus library we ported, in trilateralBazzar or mandelbrot you can find hxPixels use with Kha.

hxPixels seems to be much more optimized than mine. Will checkout those libraries you mention. I was thinking about adding support for text using bitmap fonts, and now you also given me the idea of playing a little bit with fractals ! :)  


> I have this repo for loading images a wrapper for format library and makes use of hxPixels:
> https://github.com/nanjizal/folder
> You can see it's use with
> https://github.com/nanjizal/hxRectPack2D_pixel
> For nodejs slightly different approach
> https://github.com/nanjizal/folderNode

Thanks! Probably I will grab some code and test gif and jpg that seems to partially work. 

again, I'm pretty new with this technology and my first impression was that the community was not so active but I see is not true. Thanks again,

**Comment by nanjizal** [comment]
I am working away from software the summer but haxe discord and gitter channels are quite active and there is a community forum you can find on the haxe site.
